RELEASE/scripts/sl_ascend/sl_cooking.ash

```diff
@@ -1582,37 +1592,163 @@ void consumeStuff()
 	}
 }
 
-boolean loadConsumables(string _type, item[int] item_backmap, int[int] cafe_backmap, float[int] adv, int[int] space)
+int SL_ORGAN_STOMACH = 1;
+int SL_ORGAN_LIVER   = 2;
+
+int SL_OBTAIN_NULL  = 100;
+int SL_OBTAIN_CRAFT = 101;
+int SL_OBTAIN_PULL  = 102;
+
+// Used internally for knapsack optimization.
+record ConsumeAction
+{
+	// exactly one of these is non-none
+	item it;
+	int cafeId;
+
+	int size;           // how much of organ is used
+	float adventures;   // expected adv from (thing)
+
+	float desirability; // adv count that will be used for optimization
+	                    // (lower for pulls, higher for buffs/tower keys)
+
+	int organ;          // SL_ORGAN_*
+	int howToGet;       // SL_OBTAIN_*
+};
+
+string consumable_name(ConsumeAction action)
+{
+	string name = "<name not found>";
+	if (action.it != $item[none]) name = to_string(action.it);
+	else if (action.organ == SL_ORGAN_LIVER) name = cafeDrinkName(action.cafeId);
+	else if (action.organ == SL_ORGAN_STOMACH) name = cafeFoodName(action.cafeId);
+	return name;
+}
+
+string to_string(ConsumeAction action)
+{
+	string ret = "";
+	ret += "ConsumeAction(it="+action.it;
+	ret += ",cafeId="+action.cafeId;
+	ret += ",size="+action.size;
+	ret += ",adventures="+action.adventures;
+	ret += ",desirability="+action.desirability;
+	ret += ",organ="+action.organ;
+	ret += ",howToGet="+action.howToGet;
+	ret += ")";
+	return ret;
+}
+
+ConsumeAction MakeConsumeAction(item it)
+{
+	int organ = it.inebriety > 0 ? SL_ORGAN_LIVER : SL_ORGAN_STOMACH;
+	int size = max(it.inebriety, it.fullness);
+	float adv = expectedAdventuresFrom(it);
+	return new ConsumeAction(it, 0, size, adv, adv, organ, SL_OBTAIN_NULL);
+}
+
+boolean slPrepConsume(ConsumeAction action)
+{
+	print(to_string(action));
+	if(action.howToGet == SL_OBTAIN_PULL)
+	{
+		print("slPrepConsume: Pulling a " + action.it, "blue");
+		action.howToGet = SL_OBTAIN_NULL;
+		return pullXWhenHaveY(action.it, 1, item_amount(action.it));
+	}
+	else if(action.howToGet == SL_OBTAIN_CRAFT)
+	{
+		print("slPrepConsume: Crafting a " + action.it, "blue");
+		action.howToGet = SL_OBTAIN_NULL;
+		return create(1, action.it);
+	}
+	else if (action.howToGet == SL_OBTAIN_NULL)
+	{
+		print("slPrepConsume: Doing nothing to get a " + action.it, "blue");
+	}
+	return true;
+}
+
+boolean slConsume(ConsumeAction action)
+{
+	if (action.howToGet == SL_OBTAIN_PULL || action.howToGet == SL_OBTAIN_CRAFT)
```

Comment:
It'd be simpler to just check `if(action.howToGet != SL_OBTAIN_NULL)`, not that it makes a big difference either way.